### PR TITLE
feat(gate): reject unjustified coverage expansion before fixer dispatch

### LINKED ISSUE #2032: Reject unjustified coverage expansion before fixer dispatch
## Context

Repeated review rounds can turn a narrowly scoped fix into disproportionate test growth. Coverage requests currently lack a portable admission rule: they may ask for equivalent permutations at helper, resolver, CLI, generated-mirror, or harness-labelled layers without identifying a distinct missing behavior. The existing judge can reject findings before fixer dispatch, but its contract does not require unjustified coverage expansion to use that path.

This affects dev-loops itself and the packaged guidance consumed through Pi, Claude Code, and Codex. The primary beneficiary is the operator running dev-loop, who needs demonstrated defects preserved while unsupported test expansion stops before another fixer round.

## Decided approach

`skills/docs/validation-policy.md` remains the single canonical owner of test-evidence economy. A coverage request is actionable only when it identifies the observable behavior or risk, explains why existing evidence misses it, and selects the cheapest authoritative seam. One layer is authoritative by default; another layer is justified only by a distinct boundary failure such as argument parsing, transport, serialization, packaging, or actual harness routing.

The existing judge disposition path enforces the rule: a coverage finding that fails the admission rule is `reject`, so the existing act-list filter keeps it out of fixer dispatch. `scopeDrift` remains supporting evidence rather than a new gate. Demonstrated fail-open, security, acceptance-criterion, or public-contract defects remain actionable. Justified expansion that changes the approved scope continues through the existing spec-authority/operator-approval path.

The existing ≥90% coverage target is diagnostic, not sufficient grounds for a finding. Material growth—an extra test layer without a distinct boundary risk, repeated permutations without a distinct failure mechanism, or work beyond recorded verification scope—must still satisfy the admission rule. Existing size-budget signals may support judgment but are not a test-to-production ratio.

## AC / DoD matrix

| Acceptance-criterion outcome | Required completion evidence |
|---|---|
| A single portable rule in `skills/docs/validation-policy.md` makes every coverage request name the protected observable behavior or risk, the insufficiency of existing evidence, and the cheapest authoritative seam. Percentage-only, missing-test-only, equivalent-permutation, generated-mirror-only, and harness-label-only demands are non-actionable; the existing coverage percentage is diagnostic only. | One stable rule block is added to the canonical policy. Reviewer and judge guidance references that owner without restating it; generated/package reproducibility and link checks pass. |
| One authoritative test layer is the default. Additional unit, integration, CLI, package, or harness coverage is allowed only for a distinct boundary failure. Representative checks may cover equivalent error paths, while independent uncertainty mechanisms capable of unsafe acceptance remain covered. | A compact review rehearsal rejects an already-covered permutation, accepts a missing public-boundary behavior, and retains a demonstrated fail-open uncertainty defect. Existing useful boundary and fail-closed coverage remains green. |
| The existing judge evaluates coverage justification before fixer dispatch. A failing coverage request receives `reject`; the existing act-list filter dispatches only accepted findings. Demonstrated defects retain their established disposition rules. | Existing `runJudgePass`/disposition evidence proves `reject` cannot enter the act list. One compact judge rehearsal rejects an already-covered permutation, accepts a missing public-boundary behavior, and retains a demonstrated fail-open defect; no new dispatch code is added. |
| A repeated coverage demand remains rejected absent new evidence naming a distinct behavior or boundary risk. Prior-round judge ledgers support the comparison without a new durable rejection registry. | One two-round rehearsal supplies the prior rejection to the existing judge context and proves the unchanged demand remains rejected while genuinely new evidence is reconsidered. |
| Pi, Claude Code, Codex, and installed consumers receive equivalent behavior from their actual supported canonical or generated path; inert counterpart paths may rely on unchanged passing evidence. | Existing cross-harness and clean-package verification proves the policy and referenced roles resolve through supported delivery paths. New behavioral scenarios are not multiplied across harness, layer, and malformed-input matrices unless a distinct wiring boundary changed. `bun run verify` passes on the pinned toolchain. |

## Non-goals

- Cleaning up the existing test suite; that is tracked by #2033 and its children.
- A semantic test deduplicator, per-test behavior or rejection registry, new reviewer, new dispatch gate/service, fixed test-to-production ratio, or universal test-count threshold.
- Weakening fail-closed behavior, required public contracts, immutable accepted acceptance criteria, or cross-harness non-regression.
- Expanding supported malformed-input semantics merely to create test cases.
- Reviewer-count scaling (#1984), merge-size enforcement (#1960), or broad contract compression (#2035 and #2036).

## Delivery line

Implement only on the isolated `1.0.2-slim` integration line created from the exact v1.0.2 release baseline. The PR targets that line, not `main`, and includes no unrelated feature work.

## Scope boundary

This issue owns the canonical coverage-admission rule (validation-policy.md) and the judge reject path that blocks unjustified coverage expansion before fixer dispatch. It does NOT own the existing test-suite slimming stream (#2033) or contract/prompt compression (#2035, #2036).

### PR DESCRIPTION
**Pull request by mfittko**
## Summary

Adds a single portable coverage-request admission rule so unjustified coverage expansion is rejected before fixer dispatch, while demonstrated defects are preserved. `skills/docs/validation-policy.md` remains the single canonical owner of test-evidence economy: a new `VALIDATE-COVERAGE-ADMISSION` rule block makes a coverage request actionable only when it names the protected observable behavior or risk, why existing evidence misses it, and the cheapest authoritative seam. The existing judge disposition path enforces it — a coverage finding that fails admission is `reject`, so the existing act-list filter keeps it out of fixer dispatch. No new dispatch code, reviewer, gate, service, or rejection registry; `agents/judge.agent.md` and `agents/review.agent.md` reference the owner without restating it; `.claude` mirrors regenerate in lockstep.

## Acceptance criteria

- [x] A single portable rule in `skills/docs/validation-policy.md` makes every coverage request name the protected observable behavior or risk, the insufficiency of existing evidence, and the cheapest authoritative seam; percentage-only, missing-test-only, equivalent-permutation, generated-mirror-only, and harness-label-only demands are non-actionable; the existing coverage percentage is diagnostic only.
- [x] One authoritative test layer is the default; additional unit, integration, CLI, package, or harness coverage is allowed only for a distinct boundary failure; representative checks may cover equivalent error paths while independent uncertainty mechanisms capable of unsafe acceptance remain covered.
- [x] The existing judge evaluates coverage justification before fixer dispatch; a failing coverage request receives `reject` and the existing act-list filter dispatches only accepted findings; demonstrated defects retain their established disposition rules.
- [x] A repeated coverage demand remains rejected absent new evidence naming a distinct behavior or boundary risk; prior-round judge ledgers support the comparison without a new durable rejection registry.
- [x] Pi, Claude Code, Codex, and installed consumers receive equivalent behavior from their actual supported canonical or generated path; inert counterpart paths may rely on unchanged passing evidence.

## Definition of done

- [x] One stable rule block added to the canonical policy; reviewer and judge guidance reference that owner without restating it; generated/package reproducibility and link checks pass.
- [x] Compact judge/review rehearsals reject an already-covered permutation, accept a missing public-boundary behavior, and retain a demonstrated fail-open defect; existing boundary and fail-closed coverage remains green.
- [x] Existing `runJudgePass`/disposition evidence proves `reject` cannot enter the act list; no new dispatch code is added.
- [x] A two-round rehearsal proves the unchanged demand stays rejected (matched by the existing finding fingerprint) while genuinely new distinct-boundary evidence is reconsidered.
- [x] Cross-harness and clean-package verification resolves the policy and referenced roles through supported delivery paths; `bun run verify` passes on the pinned toolchain.

## Non-goals

- Cleaning up the existing test suite (tracked separately).
- A semantic test deduplicator, per-test or rejection registry, new reviewer, new dispatch gate/service, fixed test-to-production ratio, or universal test-count threshold.
- Weakening fail-closed behavior, required public contracts, immutable accepted acceptance criteria, or cross-harness non-regression.
- Expanding supported malformed-input semantics merely to create test cases.

## Delivery line

Targets the isolated `1.0.2-slim` integration line, not `main`. No unrelated feature work.

Closes #2032

### CODE REVIEW

**Review by mfittko — COMMENTED** [review]
### Gate review: `draft_gate`


**Reviewed head SHA:** `fad1889d0550a02a517b1493372596f4fab80b44`
**Verdict:** clean
**Execution mode:** fanout_fanin

**Findings summary:** 2 angles reviewed; 2 findings (see per-angle breakdown below).

**Inline findings:** 1 (⚪ 1 nit) — on the diff, see Files changed. Angles: coverage.

Body-only findings — no anchorable changed line, so carried in full here (plain list, angle in brackets, `file:line` linked to the blob when known):

- ⚪ nit — PR body's Acceptance-criteria table carries an 'Evidence' column and the Definition-of-done section asserts present-tense results ('checks pass', 'bun run verify passes (8221 pass, 0 fail)'). Per review.agent.md, verdict/evidence content belongs in the review verdict, not the PR description. The AC/DoD/Non-goals themselves are correct and self-contained; only the embedded pass-claims are verdict-flavored. — _`deferred`_ — judge: _`reject`_ _(pr-description)_


**Next action:** Draft gate clean (2 nits rejected with rationale, no head-bump). Ready for review + Copilot when authorized.

File: test/loop/judge-pass.test.mjs [hunk 1]
```diff
@@ -363,6 +363,69 @@ test("judgePassCli reject records a fingerprint audit entry and creates no issue
...
+        { index: 0, disposition: "reject", rationale: "no new evidence since the prior-round rejection; still percentage-only", criterion: "VALIDATE-COVERAGE-ADMISSION" },
+        { index: 1, disposition: "act", rationale: "distinct serialization boundary, names the protected behavior", criterion: "VALIDATE-COVERAGE-ADMISSION" },
+      ],
+    }),
+    HEAD,
+  );
+  assert.deepEqual(round2.act.map((f) => f.summary), ["serialization boundary emits malformed JSON on NaN — uncovered"]);
+  // The unchanged demand is identified by the existing fingerprint mechanism —
+  // no new rejection registry is introduced.
+  assert.equal(fingerprintFinding(coverageDemand), fingerprintFinding({ ...coverageDemand }));
```
Comment (mfittko) [inline test/loop/judge-pass.test.mjs]: **nit** (`coverage`): Two-round rehearsal's closing assertion is tautological: fingerprintFinding(coverageDemand) === fingerprintFinding({ ...coverageDemand }) hashes the same literal twice (a shallow copy), proving only that fingerprintFinding is deterministic on identical input. It does not demonstrate AC-4's named mechanism 'the unchanged demand is identified by the existing fingerprint' across the two rounds. The reject-persistence behavior itself IS covered by the act-list assertions (round1.act.length === 0; round2.act contains only the new boundary finding), so this is the decorative clause, not the protected behavior. — judge: reject

Comment (mfittko) [reply test/loop/judge-pass.test.mjs]: Resolved at gate close (round 1, fingerprint c99c3b4174e6f0e1, severity nit, angle coverage): Examined on merits: "Two-round rehearsal's closing assertion is tautological: fingerprintFinding(coverageDemand) === fingerprintFinding({ ...coverageDemand }) hashes the same literal twice (a shallow copy), proving only that fingerprintFinding is deterministic on identical input. It does not demonstrate AC-4's named mechanism 'the unchanged demand is identified by the existing fingerprint' across the two rounds. The reject-persistence behavior itself IS covered by the act-list assertions (round1.act.length === 0; round2.act contains only the new boundary finding), so this is the decorative clause, not the protected behavior." was reviewed against current scope and acceptance criteria; it is cosmetic and does not warrant a fixer cycle or tracked follow-up. nit findings are resolved with rationale at gate close, with no fixer cycle and no tracked follow-up issue (net-reduction disposition policy).

File: test/loop/judge-pass.test.mjs [hunk 2]
```diff
@@ -363,6 +363,69 @@ test("judgePassCli reject records a fingerprint audit entry and creates no issue
...
+      { index: 1, disposition: "act", rationale: "distinct argument-parsing boundary, cheapest authoritative seam", criterion: "VALIDATE-COVERAGE-ADMISSION" },
+      { index: 2, disposition: "act", rationale: "demonstrated fail-open uncertainty defect remains actionable" },
+    ],
+  });
+  const result = runJudgePass(findings, v, HEAD);
+  assert.deepEqual(result.act.map((f) => f.summary), [
+    "no test for the CLI argument-parsing boundary (public seam)",
+    "resolver accepts unverifiable input (fail-open)",
+  ]);
+  assert.ok(!result.act.some((f) => f.judgeDisposition === "reject"), "a rejected coverage permutation never enters the act list");
```
[the remark this answers is not in the document]
Comment (mfittko) [reply test/loop/judge-pass.test.mjs]: Correct observation, dispositioned as a nit (resolve-with-rationale, no head-bump), per the net-reduction bar.

The redundant `!result.act.some(judgeDisposition === "reject")` guard and the tautological cross-round fingerprint equality do not carry the acceptance criterion: the protected behavior — a rejected coverage demand is excluded from the fixer act list while genuinely new distinct-boundary evidence is acted on — is already proven by the meaningful `assert.deepEqual(result.act.map(f => f.summary), ...)` assertions in both rounds (round 1 act is empty; round 2 act is exactly the new boundary finding). Those flagged lines are decorative noise on top of assertions that already hold.

Leaving them is a cosmetic test-quality nit, not a correctness or AC gap, so it is resolved in-thread rather than fixed here — a head-bump would trigger a full draft + pre-approval + Copilot re-gate cascade for a passing, non-load-bearing assertion. A future test-suite tidy pass (tracked by the existing suite-slimming stream) can drop the redundant lines or convert them to the subset/enriched-entry assertions you suggest.

File: agents/review.agent.md [hunk 3]
```diff
@@ -84,11 +84,12 @@ GATE-EXEC-PRIME primer-then-parallel pattern.
 - Definition of done coverage: are verification, documentation, CI, release, and operational expectations fully met?
 - Non-goals discipline: does the change avoid introducing or silently shipping work outside the stated scope?
 - Coding best practices: prefer KISS, SRP, YAGNI, readability, maintainability, and coherent test coverage.
+- Coverage findings follow [VALIDATE-COVERAGE-ADMISSION](../skills/docs/validation-policy.md) (canonical owner; the non-actionable classes live there — do not restate them here): raise a coverage gap only when you can name the protected behavior, the insufficiency of the existing evidence, and the cheapest authoritative seam; otherwise leave it unraised.
```
Comment (mfittko) [inline agents/review.agent.md:87]: **nit** (`pr-checklist`): DoD1's checked box claims reviewer guidance references the owner 'without restating it', but this line restates the canonical rule's positive three-part admission test (name the protected behavior, the insufficiency of existing evidence, the cheapest authoritative seam) verbatim in substance. The substantive non-actionable classes and disposition mechanics are correctly delegated to the owner, so this is a brief point-of-use actionable cue rather than material duplication; the DoD box is not dishonestly ticked. — judge: reject

**Review by mfittko — COMMENTED** [review]
### Gate review: `pre_approval_gate`


**Reviewed head SHA:** `f4c1e070b9b7a12acc5cb8acf63bc7498c306b69`
**Verdict:** clean
**Execution mode:** fanout_fanin

**Findings summary:** 7 angles reviewed; 1 finding (see per-angle breakdown below).

**Inline findings:** 1 (⚪ 1 nit) — on the diff, see Files changed. Angles: pr-checklist.

**Clean (6):** docs, yagni, contradiction-lens, acceptance-criteria, dry, renderer-security

**Gate evidence note:** Copilot review rounds exhausted (1/1); current head has zero unresolved threads and green CI, so pre_approval_gate fallback is allowed without another Copilot re-request.

**Next action:** Pre-approval gate clean (1 nit rejected with rationale, no head-bump). Merge-ready pending orchestrator.

File: agents/review.agent.md (hunk 3 above)
Comment (mfittko) [reply agents/review.agent.md:87]: Resolved at gate close (round 1, fingerprint 12ade86ede758a4e, severity nit, angle pr-checklist): Examined on merits: "DoD1's checked box claims reviewer guidance references the owner 'without restating it', but this line restates the canonical rule's positive three-part admission test (name the protected behavior, the insufficiency of existing evidence, the cheapest authoritative seam) verbatim in substance. The substantive non-actionable classes and disposition mechanics are correctly delegated to the owner, so this is a brief point-of-use actionable cue rather than material duplication; the DoD box is not dishonestly ticked." was reviewed against current scope and acceptance criteria; it is cosmetic and does not warrant a fixer cycle or tracked follow-up. nit findings are resolved with rationale at gate close, with no fixer cycle and no tracked follow-up issue (net-reduction disposition policy).
